Add draft of PlatformIO post (#164)
memfault/interrupt: _drafts/platformio.md

> title: What About PlatformIO?

Reviewer: ```suggestion
---
title: What About PlatformIO?
```

title: What About PlatformIO?
description:
    An overview of PlatformIO, as well as a getting started example on STM32.
author: francois
---

> A few weeks ago, I wrote about MCU SDKs and how frustrating I sometimes found it

Reviewer: ```suggestion
A few weeks ago, [I wrote about MCU SDKs]({% post_url 2020-11-10-the-best-and-worst-mcu-sdks %}) and how frustrating I sometimes found it
```

FreeRTOS, true frameworks such as Arduino, and alternative APIs such as
libopencm3 and CMSIS.
3. The "board" which defines the specific board you are using. PlatformIO comes
   with hundreds of board configuration for popular development boards like the
STM32 Discovery boards.

The headline numbers are impressive: 41 platforms, 23 frameworks, and 924 boards
as of this writing.

However, reality is a bit less rosy: while some MCU families are extremely well
covered, others have very little support.

One important example is official vendor SDK. PlatformIO only supports the
vendor SDKs for the following MCUs, which it considers individual "frameworks":
* Espressif family (ESP8266, ESP32)
* Kendryte K210
* STM32
* Gigadevice GD32V
* SiFive FE310 and FU540

If you are able to piggy back on an RTOS, things look a bit better. For example,
you can create a Zephyr project with PlatformIO for Atmel SAM, Freescale
Kinetis, Nordic nRF5, NXP iMX RT, NXP LPC, Silabs EFM32, and STM32.

You can use the PlatformIO client to list supported frameworks, platforms, and
boards. The list changes regularly, as the project is very active. Here's the
command you want to run:

> ```

Reviewer: Put new lines between each command.

Author: 👍

# List platforms
$ pio platform search
# List frameworks
$ pio platform frameworks
# List boards
$ pio boards
```

### Package Management

> PlatformIO hosts a registry of embedded library which can be added to every

Reviewer: ```suggestion
PlatformIO hosts a registry of embedded libraries which can be added to every
```

...
```

Installing them is done with the `pio install` command.

### Build System

PlatformIO is an opinionated system, and as long as you follow its rules
building a project with it is seamless. Simply add your project source under
`src`, your header files under `include`, and your private libraries under
`lib`. Run `pio run`, and you're cooking with gas!

Under the hood, it uses the SCons[^scons] build system. Most of the
time, this is of no consequence to the user as plenty of configuration options
are exposed via the `platformio.ini` file used to configure each project.

For example, you can change compiler and linker flags by adding the
`build_flags` variable to your `platformio.ini`:

```
[env:release]
platform = ststm32
board = disco_f429zi
framework = stm32cube
build_flags = -Wall -Wl,--gc-sections
```

Should you need to customize your build beyond what the `.ini` file provides,
you will need to write SCons scripts which are invoked by the PlatformIO build
environment. SCons scripts are written in Python.

Two hooks are available for those scripts: a PRE hook which runs before the
build executes, a POST hook which runs after the build. You can use these hooks

> to chaange the build instructions for individual files, pre-process some source

Reviewer: ```suggestion
to change the build instructions for individual files, pre-process some source
```

code, or post-process the build artifacts[^advanced_scripting].

> While some may chafe at the need to learn yet another build system - are for one

Reviewer: ```suggestion
While some may chafe at the need to learn yet another build system - I for one
```

still like my Makefiles - SCons is a mature and powerful build system.

## Getting Started with the PlatformIO CLI

> My favorite way to use PlatformIO is the CLI, also known as "PlatformIO Core".

Reviewer: ```suggestion
My favorite way to use PlatformIO is with the CLI, also known as "PlatformIO Core".
```

Installing the PlatformIO CLI is easy. On some platform, you can use your
package manager. For example, on mac:
```
$ brew install platformio
```

For all other platforms, you can use the `get-platformio.py` script, available
on their website[^get-platformio]. Download it and run it with `python
get-platformio.py`.

At this point, you should have `platformio` available as a CLI utility, as well
as a `pio` alias for faster use.

### Creating a project

The `pio` tool comes with project management commands. Let's say we want to
start a new project for our STM32F429zi discovery board.

First, we'll create a folder:
```
$ mkdir my-project
$ cd my-project
```

Then, we'll initialize the project with `pio project init`, and pass it a board
and a framework

```
$ pio project init --board disco_f429zi --project-options "framework=STM32Cube"
...
Project has been successfully initialized! Useful commands:
`pio run` - process/build project from the current directory
`pio run --target upload` or `pio run -t upload` - upload firmware to a target
`pio run --target clean` - clean project (remove compiled files)
`pio run --help` - additional information
```

### Toolchains and linker scripts

> On many of my project, I use a specific version of the GCC toolchain, and to

Reviewer: ```suggestion
On many of my projects, I use a specific version of the GCC toolchain, and to
```

 - tool-ldscripts-ststm32 0.1.0
 - toolchain-gccarmnoneeabi 1.40804.0 (4.8.4)
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 48 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Compiling .pio/build/release/src/main.o
...
Linking .pio/build/release/firmware.elf
Checking size .pio/build/release/firmware.elf
Building .pio/build/release/firmware.bin
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [          ]   0.0% (used 44 bytes from 262144 bytes)
Flash: [          ]   0.1% (used 1296 bytes from 2097152 bytes)
======================================= [SUCCESS] Took 10.92 seconds =====================================
```

Here's everything PlatformIO does for us here:

1. Download the library and framework files (e.g. chip SDK, RTOS, ...) for our project
2. Build all C, C++, and assembly files in our `src` folder
3. Build all the private libraries in `lib` and create static libraries for them
4. Link everything into an ELF
5. Copy it into a BIN
6. Print out RAM and Flash usage

All with minimal configuration!

### Unit Tests

> Regular readers of Interrupt will know that we are [big fans of unit
> testing XXX TODO](). PlatformIO has built in supports for running tests on your

Reviewer: ```suggestion
Regular readers of Interrupt will know that we are [big fans of unit
testing]({% post_url 2019-10-08-unit-testing-basics %}). PlatformIO has built in supports for running tests on your
```
